File: courseworks/coursework-one-kotlin/src/sml/instructions/InstructionFactory.java
package sml.instructions;

import sml.Instruction;

import java.lang.reflect.InvocationTargetException;

public class InstructionFactory {

  /**
   * Generates the required Instruction
   * @param opcode Operation of the language
   * @param args the arguments for the operation
   * @return the required Instruction for the SML Machine
   */
  public static Instruction generateInstruction(String opcode, String[] args) {

    Class myClass = classProducer(opcode);
    Object result = null;
    try {

      Class[] typeOfRequiredArguments = myClass.getConstructors()[0].getParameterTypes();

      Object[] requiredArgs = constructorArguments(typeOfRequiredArguments, args);
      result = myClass.getConstructors()[0].newInstance(requiredArgs);
    } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
      e.printStackTrace();
    }
    return (Instruction)result;
  }

  /**
   * Creates an array of objects
   * @param typeOfRequiredArgument an array with the class of each argument required by the constructor
   * @param inputArgs an array of strings with the input of the user
   * @return an array of objects which will be passed the to constructor
   */
  private static Object[] constructorArguments(Class[] typeOfRequiredArgument, String[] inputArgs) {

    Object[] args = new Object[typeOfRequiredArgument.length];

    if(inputArgs.length != typeOfRequiredArgument.length) {
      throw new IllegalArgumentException("Wrong number of Arguments");
    }

    for(int i = 0 ; i < typeOfRequiredArgument.length ; i ++) {

      if(typeOfRequiredArgument[i].getSimpleName().equals("int")) {
        args[i] = Integer.parseInt(inputArgs[i]);
      } else {
        args[i] = inputArgs[i];
      }
    }
    return args;
  }

  /**
   * Capitalize the first character of the label and concatenates it with "Instruction"
   * @param label Operation of the Language
   * @return the required Class
   */
  private static Class classProducer(String label) {
    Class myClass= null;
    try {
      myClass = Class.forName("sml.instructions."+label.substring(0,1).toUpperCase() + label.substring(1)+"Instruction");
    } catch (ClassNotFoundException e) {

      e.printStackTrace();
    }

    return myClass;
  }
}

File: courseworks/coursework-one-kotlin/src/sml/instructions/InstructionFactory.kt
package sml.instructions

import sml.Instruction
import kotlin.reflect.KClass
import kotlin.reflect.KFunction
import kotlin.reflect.KParameter
import kotlin.reflect.full.createType
import kotlin.system.exitProcess


/**
 * InstructionFactory Class
 * @param instruction Instruction of the language
 * @param args the arguments for the operation
 */
class InstructionFactory(private val instruction: String, private val args: Array<String> ) {

    /**
     * Generates the required Instruction
     * @return the required Instruction for the SML Machine
     */
    fun generateInstruction() : Instruction {

        val myCLass = generateClass()
        if(myCLass!= null) {
           val myConstructor:KFunction<Any> = myCLass.constructors.elementAt(0)
           val requiredParameters = myConstructor.parameters

           val k =  constructorArguments(requiredParameters, args)
           return myConstructor.call(*k) as Instruction
        }

        return NoOpInstruction(args[0], "Unknown instruction [$instruction]" )



    }

    /**
     * Generates the required class by formatting the instruction string that was given to the class
     * and calling Class.forName function
     * @return the required Class
     */
    private fun generateClass(): KClass<out Any>? {
        var kClass :KClass<out Any>? = null
        try {
            kClass = Class.forName("sml.instructions."
                    + instruction.substring(0,1).toUpperCase() + instruction.substring(1) + "Instruction").kotlin

        } catch (e: ClassNotFoundException) {

        }
        return kClass
    }

    /**
     * Creates an array with required arguments from the constructor
     * @param requiredParameters an List with the Kparameters required by the constructor
     * @param inputArray an array of strings with the input of the user
     * @return an array which will be passed the to constructor
     */
    private fun constructorArguments(requiredParameters: List<KParameter>, inputArray: Array<String>): Array<Any> {

        // Check if the size
        if(requiredParameters.size != inputArray.size) {
            println(if(requiredParameters.size > inputArray.size) "Insufficient arguments for $instruction" else
                "Abundant arguments for $instruction."
                    + "\nRequired ${requiredParameters.size} found ${inputArray.size}")
            exitProcess(-1)
        }

        val myArray = Array<Any>(requiredParameters.size,{1}) //creating an array with size equal to the constructors
        // argument, with initial values set to 1
        var i = 0
        for(e in requiredParameters) {
            when {

                e.type == Int :: class.createType() -> {
                    // if the required type is int then we cast
                    // the given value to int and we add it to the array we are going to return
                    myArray[i] = inputArray[i].toInt()
                    i+=1
                }
                e.type == String :: class.createType() -> {

                    myArray[i] = inputArray[i]
                    i+=1
                }

            }

        }
        return myArray

    }

}



